metrics/internal/collectors/storage-cluster.go

```diff
@@ -0,0 +1,85 @@
+package collectors
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+
+	v1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	"github.com/red-hat-storage/ocs-operator/metrics/internal/options"
+)
+
+type StorageClusterCollector struct {
+	KMSServerConnectionStatus *prometheus.Desc
+	Informer                  cache.SharedIndexInformer
+}
+
+var _ prometheus.Collector = &StorageClusterCollector{}
+
+func NewStorageClusterCollector(opts *options.Options) *StorageClusterCollector {
+	cl, err := GetOcsClient(opts)
+	if err != nil {
+		klog.Errorf("Unable to get client: %v", err)
+		return nil
+	}
+	lw := cache.NewListWatchFromClient(cl, "storageclusters", metav1.NamespaceAll, fields.Everything())
+	sharedIndexInformer := cache.NewSharedIndexInformer(lw, &v1.StorageCluster{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	return &StorageClusterCollector{
+		KMSServerConnectionStatus: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "storagecluster", "kms_connection_status"),
+			`KMS Connection Status; 0: Connected, 1: Failure.`,
+			[]string{"name", "namespace"},
+			nil,
+		),
+		Informer: sharedIndexInformer,
+	}
+}
+
+func (c *StorageClusterCollector) Run(stopCh <-chan struct{}) {
+	go c.Informer.Run(stopCh)
+}
+
+func (c *StorageClusterCollector) Describe(ch chan<- *prometheus.Desc) {
+	ds := []*prometheus.Desc{
+		c.KMSServerConnectionStatus,
+	}
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+func (c *StorageClusterCollector) Collect(ch chan<- prometheus.Metric) {
+	scLister := NewStorageClusterLister(c.Informer.GetIndexer())
+	storageClusters := getAllStorageClusters(scLister)
+	if len(storageClusters) > 0 {
+		c.collectKMSConnectionStatuses(ch, storageClusters)
+	}
+}
+
+func getAllStorageClusters(lister StorageClusterLister) []*v1.StorageCluster {
+	storageClusters, err := lister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("couldn't list StorageCluster: %v", err)
+		return nil
+	}
+	return storageClusters
+}
+
+func (c *StorageClusterCollector) collectKMSConnectionStatuses(ch chan<- prometheus.Metric, storageClusters []*v1.StorageCluster) {
+	for _, storageCluster := range storageClusters {
+		v := 0
+		if storageCluster.Status.KMSServerConnection != *new(v1.KMSServerConnectionStatus) && storageCluster.Status.KMSServerConnection.KMSServerConnectionError != "" {
```

Comment:
As `Status.KMSServerConnection` is not a pointer, we only have to check the field `Status.KMSServerConnection.KMSServerAddress` is populated (and may be an extra check of a syntactically valid address) or not, and there is no error message (in `KMSServerConnectionError` field).

```suggestion
		if storageCluster.Status.KMSServerConnection.KMSServerAddress != "" && storageCluster.Status.KMSServerConnection.KMSServerConnectionError == "" {
```

or even better

```suggestion
		if validKMSAddress(storageCluster.Status.KMSServerConnection.KMSServerAddress) && storageCluster.Status.KMSServerConnection.KMSServerConnectionError == "" {
```
where `validKMSAddress()` will check the given argument is a valid address (or not), but this can be a future enhancement/check.

Reply:
Opting for the first approach for now.
Though, I was wondering since we are making it more granular to check if KMS is actually enabled or not, should we add a third status named `Not Connected`  in the collector?
Also, I am not exactly sure if we need to add a `prometheusAlerts` for the `kms_connection_status`? Isn't that the job of the collector to pass such alerts to prometheus if the validation fails?

Sorry, I am a bit new to the code and the docs isn't much helpful in this case either so a bit confused.

Reply:
> Anything other than Connected status, that is: other than the metric value ZERO, should be not connected.
Another improvement here is we can have a value of -1 (for the metric) which can refer to KMS not enabled status (but again may not be the current scope).

Currently, I have added 0: Connected, 1: NotConnected, 2: Failure.
Are you suggesting we should probably, update it to `-1: NotConnected, 0: Connected, 1: Failure` ?

Reply:
Nope, the numbers are fine. Added comment in the below review (as I have seen that changes a bit late). Please check